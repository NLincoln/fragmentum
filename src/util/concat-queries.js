export const concatQueries = (arr, joinStr = ", ") => {
  const reduced = arr.filter(f => f.query).reduce(
    (prev, curr) => ({
      query: prev.query.concat(curr.query),
      binds: prev.binds.concat(curr.binds)
    }),
    {
      query: [],
      binds: []
    }
  );
  return {
    query: reduced.query.join(joinStr),
    binds: reduced.binds
  };
};
