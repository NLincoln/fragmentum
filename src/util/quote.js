export default function quote(val, opts = {}) {
  if (val === "*") {
    return val;
  }
  if (opts.single) {
    return `'${val.replace(`'`, `\\'`)}'`;
  }
  if (opts.parens) {
    return `(${val})`;
  }
  return `"${val.replace(`"`, `\\"`)}"`;
}
