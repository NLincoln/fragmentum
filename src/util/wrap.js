export default Klass => (...args) => new Klass(...args);
