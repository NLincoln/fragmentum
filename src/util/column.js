import quote from "./quote";

export default raw =>
  raw
    .split(".")
    .map(quote)
    .join(".");
