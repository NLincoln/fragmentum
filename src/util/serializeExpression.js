import { Builder } from "../builder";
import Expression from "../expressions/expression";
import FromFragment from "../fragments/from";
import quote from "./quote";
import columnQuote from "./column-quote";

export const serializePair = ([key, alias]) => {
  if (key && alias) {
    return `${columnQuote(key)} AS ${quote(alias)}`;
  }
  return null;
};
const serializeObject = column =>
  Object.keys(column)
    .map(key => serializePair([key, column[key]]))
    .join(", ");

const serializeColumn = column => {
  if (Array.isArray(column)) {
    return serializePair(column);
  } else if (typeof column === "object") {
    return serializeObject(column);
  }
  return columnQuote(column);
};

export const serializeMaybeExpression = column => {
  // Expressions know how to serialize themselves, FromFragment indicates that
  // we have a wrapped subquery, so we just forward that too.
  if (column instanceof Expression || column instanceof FromFragment) {
    return column.serialize();
  } else if (column instanceof Builder) {
    let { query, binds } = column.serialize({ partial: true });
    return {
      query: `(${query}) AS ${quote(column.alias)}`,
      binds
    };
  }
  return {
    query: serializeColumn(column),
    binds: []
  };
};
