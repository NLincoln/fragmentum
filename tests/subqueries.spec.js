import { builder, where, ops, bind, select, from } from "fragmentum";
import { testQuery } from "./util";

describe("subqueries", () => {
  describe("From <subquery>", () => {
    testQuery(
      "from a builder",
      builder().from(
        builder("alias")
          .select()
          .from("users")
      ),
      `FROM (SELECT * FROM "users") AS "alias"`
    );
    testQuery(
      "a garden mix of everything because I'm evil",
      builder(
        from(
          from(
            "user",
            builder("alias")
              .select()
              .from("users")
          )
        ),
        from(
          builder("alias2", where(ops.eq("user.user_id", bind("userid", 2))))
            .select()
            .from("groups")
        )
      ),
      {
        query: `FROM "user", (SELECT * FROM "users") AS "alias", (SELECT * FROM "groups" WHERE ("user"."user_id" = :userid)) AS "alias2"`,
        binds: [
          {
            userid: 2
          }
        ]
      }
    );

    testQuery(
      "from an incomplete fragment",
      builder(from(builder("alias", select()))),
      `FROM (SELECT *) AS "alias"`
    );

    testQuery(
      "from another from fragment",
      builder()
        .select("alias.user_id")
        .from(
          from(
            builder("alias", select())
              .from("users")
              .from("groups")
          )
        ),
      `SELECT "alias"."user_id" FROM (SELECT * FROM "users", "groups") AS "alias";`
    );
  });
  describe("subquery in expression", () => {
    testQuery(
      "subquery inside of expression: full test",
      () =>
        builder(
          select(),
          from("users"),
          where(
            ops.eq(
              "user_id",
              builder(
                "alias",
                select("user_id"),
                from("users"),
                where(ops.eq("group_id", bind("user_id", 2)))
              )
            )
          )
        ),
      `SELECT * FROM "users" WHERE ("user_id" = (SELECT "user_id" FROM "users" WHERE ("group_id" = :user_id)));`
    );
  });
});
