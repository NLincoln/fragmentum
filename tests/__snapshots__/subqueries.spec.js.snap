// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subqueries not providing a name 1`] = `
Object {
  "binds": Object {},
  "query": "FROM (SELECT * FROM \\"users\\") AS \\"BUILDER_anonymous0\\"",
}
`;
